### PR TITLE
Approximate Bayesian Inference Using Dropout

### PR DESCRIPTION
**Pull request by patrickhop**
I've added approximate bayesian inference using dropout in the style of https://arxiv.org/pdf/1506.02142.pdf. It can be turned on during model init using the [error_bars=True] arg. There seems to be a nice little model averaging effect going on from the multiple forward passes. Performance in the approximate bayesian setting is slightly higher.

I spot checked the error bars on a few datasets. About 70% of regression labels are within 1 std of the predicted mean, which sounds about right.

I've also fixed some more inference bugs. Multitask is now also working.

I ran yapf on the examples folder, so a lot of files were changed. Please let me know if this needs to be rolled back.

### CODE REVIEW

**Comment by rbharath** [comment]
@patrickhop This is awesome!

Would you mind rolling back the yapf on the examples? I think it's good to yapf everything, that might be better for a different PR (hard to code-review with so many changes, even if most are just yapf)

**Comment by patrickhop** [comment]
rolled back.

File: AUTHORS [hunk 1]
```diff
@@ -0,0 +1,78 @@
+Alexander Rose <alexander.rose@weirdbyte.de>
```
Comment (rbharath) [inline AUTHORS]: Woud you mind removing this file? (The same information about contributors should be available through github)

File: ChangeLog [hunk 2]
```diff
@@ -0,0 +1,1762 @@
+CHANGES
```
Comment (rbharath) [inline ChangeLog]: Would you mind removing this file?

File: deepchem/models/tensorgraph/models/graph_models.py [hunk 3]
```diff
@@ -600,28 +609,47 @@ def evaluate(self, dataset, metrics, transformers=[], per_task_metrics=False):
         labels=self.my_labels,
         weights=[self.my_task_weights])
 
-  def predict_on_smiles(self, smiles, transformers):
-    max_index = len(smiles)
-    num_batches = max_index // self.batch_size
+  def bayesian_predict(self, X, transformers=[], n_passes=4, untransform=False):
```
Comment (rbharath) [inline deepchem/models/tensorgraph/models/graph_models.py]: Could you add a docstring here?

Also, instead of accepting a numpy array `X`, this should accept either a dataset object or smiles.

File: deepchem/models/tensorgraph/models/graph_models.py [hunk 4]
```diff
@@ -507,20 +511,23 @@ def build_graph(self):
...
+    dense = Dense(out_channels=128, activation_fn=tf.nn.relu, in_layers=[gp2])
     batch_norm3 = BatchNorm(in_layers=[dense])
-    gg1 = GraphGather(
+    readout = GraphGather(
         batch_size=self.batch_size,
         activation_fn=tf.nn.tanh,
         in_layers=[batch_norm3, self.degree_slice, self.membership] +
         self.deg_adjs)
 
+    if self.error_bars == True:
```
Comment (rbharath) [inline deepchem/models/tensorgraph/models/graph_models.py]: Should `self.error_bars` be set in a constructor?

File: deepchem/models/tensorgraph/models/graph_models.py [hunk 5]
```diff
@@ -600,28 +609,47 @@ def evaluate(self, dataset, metrics, transformers=[], per_task_metrics=False):
...
+    sigma = np.concatenate(sigmas, axis=0) + 0.55
+
+    if untransform:
+      mu = undo_transforms(mu, transformers)
+      for i in range(sigma.shape[1]):
+        sigma[:, i] = sigma[:, i] * transformers[0].y_stds[i]
+
+    return mu[:max_index + 1], sigma[:max_index + 1]
+
+  def predict_on_smiles(self, smiles, transformers=[], untransform=False):
```
Comment (rbharath) [inline deepchem/models/tensorgraph/models/graph_models.py]: Would you mind adding a docstring here?

File: deepchem/models/tensorgraph/tensor_graph.py [hunk 6]
```diff
@@ -280,6 +282,36 @@ def predict_proba_on_generator(self, generator, transformers=[]):
           results.append(result)
         return np.concatenate(results, axis=0)
 
+  def bayesian_predict_on_batch(self, X, transformers=[], n_passes=4):
```
Comment (rbharath) [inline deepchem/models/tensorgraph/tensor_graph.py]: Same comments about docstring and accepting a dataset object as above.

File: deepchem/models/tensorgraph/tensor_graph.py [hunk 7]
```diff
@@ -280,6 +282,36 @@ def predict_proba_on_generator(self, generator, transformers=[]):
...
+    mu = np.mean(y_, axis=2)
+    sigma = np.std(y_, axis=2)
+
+    return mu, sigma
+
+  def predict_on_smiles_batch(self,
+                              smiles,
+                              featurizer,
+                              n_tasks,
+                              transformers=[]):
```
Comment (rbharath) [inline deepchem/models/tensorgraph/tensor_graph.py]: Could you add a docstring here?

File: deepchem/models/tensorgraph/models/graph_models.py (hunk 4 above)
Comment (rbharath) [inline deepchem/models/tensorgraph/models/graph_models.py]: Also, should the `dropout_prob` be an argument instead of fixed?

File: AUTHORS (hunk 1 above)
Comment (patrickhop) [inline AUTHORS]: don't know how that got there. will remove.

thanks for the feedback. I'll make the requested changes shortly.

Comment (patrickhop) [inline AUTHORS]: removed.

File: ChangeLog (hunk 2 above)
Comment (patrickhop) [inline ChangeLog]: removed.

File: deepchem/models/tensorgraph/models/graph_models.py (hunk 4 above)
Comment (patrickhop) [inline deepchem/models/tensorgraph/models/graph_models.py]: ```self.error_bars``` is set on 482 I believe? Is there a better way to do that?

```dropout_prob``` is fixed because inverse model precision is a function of it. Right now that term is pre-computed and is validated in this configuration.

File: deepchem/models/tensorgraph/models/graph_models.py (hunk 5 above)
Comment (patrickhop) [inline deepchem/models/tensorgraph/models/graph_models.py]: done.

File: deepchem/models/tensorgraph/tensor_graph.py (hunk 6 above)
Comment (patrickhop) [inline deepchem/models/tensorgraph/tensor_graph.py]: this is meant to be a helper function for ```bayesian_predict```, which prepares the tensor slices.

File: deepchem/models/tensorgraph/tensor_graph.py (hunk 7 above)
Comment (patrickhop) [inline deepchem/models/tensorgraph/tensor_graph.py]: done.

File: deepchem/models/tensorgraph/models/graph_models.py [hunk 8]
```diff
@@ -476,6 +479,7 @@ def __init__(self, n_tasks, **kwargs):
 
     """
     self.n_tasks = n_tasks
+    self.error_bars = True if 'error_bars' in kwargs and kwargs['error_bars'] else False
```
Comment (rbharath) [inline deepchem/models/tensorgraph/models/graph_models.py]: Ah yes, missed this in previous review

**Comment by rbharath** [comment]
It looks like the only issue is yapf. Should be good to merge once yapf issues are fixed.

**Comment by patrickhop** [comment]
great. will yapf it now

**Comment by patrickhop** [comment]
@rbharath I think it's fixed.

**Comment by patrickhop** [comment]
@rbharath do you know why travis is unhappy?

**Comment by patrickhop** [comment]
pushed more yapf

**Comment by rbharath** [comment]
LGTM.

@patrickhop Would you mind adding some unit tests in a follow-up PR?
